ui: replace NVD3 tooltip with custom details tooltip

couchand opened the pull request:
This is the last major step before replacing NVD3, using our own component for the details tooltip rather than relying on NVD3 for it.

This is not quite ready to merge.  To do items:
- [x] format the values in the tooltip
- [x] convert the rest of the dashboard pages
  - [x] runtime
  - [x] sql
  - [x] storage
  - [x] replication
  - [x] queues
  - [x] slow requests
- [x] figure out the issue with the rpc errors chart
- [x] rename `hoverThing` to `detailsTooltip`
- ~[ ] split out the last commit into smaller pieces~ (probably not worth the effort)
- [x] address in-line todo comments

(no screenshot included, as it looks the same as the current version)

couchand committed:
ui: use key rather than title for hover chart

Release note: None

couchand committed:
ui: record mouse coordinates in hover state

Release note: None

couchand committed:
ui: don't bother updating hover state before data loads

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20703)
<!-- Reviewable:end -->

couchand committed:
ui: extract count value formatter

Release note: None

couchand committed:
ui: replace NVD3 tooltip with custom details tooltip

This is the last major functionality NVD3 is giving us (besides the
charts themselves).  Replacing it with our own will enable future
customization as well as make it easier to completely excise NVD3
at some point soon.

Contributes to #18620.

Release note: None

couchand committed:
ui: correctly handle data series with no datapoints

Release note: None

couchand committed:
ui: remove filtering of datapoints

Previously we were filtering the datapoints to get around an NVD3 bug in
the tooltip.  Since we have our own tooltip now, that filtering is no
longer necessary.

Release note: None

vilterp commented:
A couple of nits. Main thing is — can we remove the config duplication? Happy to chat about it in person. (You explained it before; I forget)

---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 16 of 16 files at r5, 18 of 18 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, 16 of 16 files at r10, 2 of 2 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/redux/hover.ts, line 35 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LUgLeVXwMMr60Q3Oc:-L0LUgLeVXwMMr60Q3Od:bpmu3jf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971964e11218f84ac3775519589e768b79f9c88/pkg/ui/src/redux/hover.ts#L35)):*
> ```XML
>   hoverTime: moment.Moment;
> 
>   // The page x-coordinate of the mouse.
> ```

relative to what — the top left of the chart? the top left of the page? a little more explanation here would be good.

---

*[pkg/ui/src/util/charts.ts, line 26 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LVGNNfBzxRVqKRA3k:-L0LVGNNfBzxRVqKRA3l:bf9fvf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/util/charts.ts#L26)):*
> ```XML
> };
> 
> export function isMetricsChart(chart: Chart): chart is MetricsChart {
> ```

whoah, what syntax is this? would expect it to return `boolean`

---

*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 168 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LVKl-0EYnZQcbPHd3:-L0LVKl-0EYnZQcbPHd4:b-cbb1xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L168)):*
> ```XML
>       if (currentlyHovering) {
>         hoverTime = this.props.hoverState.hoverTime;
>         thisChart = hoverChart !== this.props.chartKey;
> ```

would expect this to be `thisChart = hoverChart === this.props.chartKey`… Am I missing something?

---

*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 127 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LXAjgDh9SBDhG6MRK:-L0LXAjgDh9SBDhG6MRL:b-z39koa) ([raw file](https://github.com/cockroachdb/cockroach/blob/096096e265197d200e93f4687c7025210a7a44aa/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L127)):*
> ```XML
>     let candidate = 0;
>     let hoverTime;
>     // Find first series with a successful bisect.
> ```

nit: would sort of prefer a `for` loop over the candidates with a `break` when you find `hoverTime`, (to me, more clear that the intention is to loop an array until we find something) but not a big difference

---

*[pkg/ui/src/views/cluster/containers/detailsTooltip/index.tsx, line 140 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LWdfWAOhxzqYwxga0:-L0LWdfWAOhxzqYwxga1:b-fux43c) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/views/cluster/containers/detailsTooltip/index.tsx#L140)):*
> ```XML
>                   <td className="legend-color-guide">
>                     <svg width={9} height={9}>
>                       <circle cx={4.5} cy={4.5} r={3.5} fill={metric.color} />
> ```

tried it locally; looks good. but isn't there supposed to be a corresponding circle on the actual time series lines as well when you're hovering? Is it hard to have that at the same time as pulling this tooltip out of NVD3?

---

*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overview.tsx, line 11 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LWBlRHfR0noGJzjFm:-L0LWBlRHfR0noGJzjFn:b-uzb296) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overview.tsx#L11)):*
> ```XML
> import { GraphDashboardProps, nodeAddress, storeIDsForNode } from "./dashboardUtils";
> 
> export const charts: ChartConfig = {
> ```

oof, all this duplication hurts. Why can't we feed this nice config object into the `LineGraph` component again? (ditto other dashboards)

---

*[pkg/ui/styl/shame.styl, line 123 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LW3z8JowiLGNaAIgE:-L0LW3z8JowiLGNaAIgF:b-sqciw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/styl/shame.styl#L123)):*
> ```Stylus
>   font-size 10px
> 
> // I really, really hate NVD3.
> ```

whoah, so many lines of shame! haha

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-:-L0LXyU8NZUIyv8XwUKh:b-yb2c8w)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: fix the tooltip for per-node charts

Release note: None

couchand committed:
ui: correct logic for linked guideline color

Release note: None

couchand committed:
ui: add some words on hover coordinates comments

Release note: None

couchand commented:
Basically it's the tooltips that include markup that make it difficult.  I suppose we could just include a React component as the configuration for that, though that seems annoying in its own way.  I think this is the right way to go, but it will take more work to get there than I thought was wise to shove into this PR.

---

Review status: 13 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/ui/src/redux/hover.ts, line 35 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LUgLeVXwMMr60Q3Oc:-L0LcRgkd3OhFSgBslAs:b-dpay1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971964e11218f84ac3775519589e768b79f9c88/pkg/ui/src/redux/hover.ts#L35)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

relative to what — the top left of the chart? the top left of the page? a little more explanation here would be good.
</blockquote></details>

I thought the work "page" made it clear, but I've added a few more words.

---

*[pkg/ui/src/util/charts.ts, line 26 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LVGNNfBzxRVqKRA3k:-L0LcWPOIuVXUAFuEGR5:bds5nr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/util/charts.ts#L26)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

whoah, what syntax is this? would expect it to return `boolean`
</blockquote></details>

ha, yeah, that's some serious TS magic for modeling discriminated unions

---

*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 168 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LVKl-0EYnZQcbPHd3:-L0Lcwyc1op_99Uet29D:bq239rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L168)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

would expect this to be `thisChart = hoverChart === this.props.chartKey`… Am I missing something?
</blockquote></details>

Hmm, that does seem backwards... fixed.

---

*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 127 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LXAjgDh9SBDhG6MRK:-L0LceeVlPi03sCiSgpU:b-mai4at) ([raw file](https://github.com/cockroachdb/cockroach/blob/096096e265197d200e93f4687c7025210a7a44aa/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L127)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

nit: would sort of prefer a `for` loop over the candidates with a `break` when you find `hoverTime`, (to me, more clear that the intention is to loop an array until we find something) but not a big difference
</blockquote></details>

I guess.  I'm not a big fan of `break` unless you really need it...

I don't like how this is working more generally, but fixing it is going to require a much bigger restructuring of the way the time series are stored, so I didn't want to tackle it at the moment.

---

*[pkg/ui/src/views/cluster/containers/detailsTooltip/index.tsx, line 140 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LWdfWAOhxzqYwxga0:-L0Lcz1oW5rHpL83ALCG:b3kubj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/views/cluster/containers/detailsTooltip/index.tsx#L140)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

tried it locally; looks good. but isn't there supposed to be a corresponding circle on the actual time series lines as well when you're hovering? Is it hard to have that at the same time as pulling this tooltip out of NVD3?
</blockquote></details>

No, those dots are wrong, NVD3 is doing it wrong.  They make it harder to see the actual values that the time series takes on.  They're chartjunk.

---

*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overview.tsx, line 11 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LWBlRHfR0noGJzjFm:-L0LdB3kWPevTacpp10j:b1vtyat) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overview.tsx#L11)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

oof, all this duplication hurts. Why can't we feed this nice config object into the `LineGraph` component again? (ditto other dashboards)
</blockquote></details>

Yes, I think that's the direction we should go in, but as I mentioned above, I didn't want to try to do it all at once.  This seemed like there was enough code in it already.

---

*[pkg/ui/styl/shame.styl, line 123 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-L0LW3z8JowiLGNaAIgE:-L0LdKQ93KaCQTbxZR5X:b-x61pga) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c75a2f7a74e901209f9453953f9d54122788cf7/pkg/ui/styl/shame.styl#L123)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

whoah, so many lines of shame! haha
</blockquote></details>

Yeah, it's super annoying that this isn't configurable at all, particularly because it's chartjunk in the first place.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20703#-:-L0LbryREk5PKXZ_nzRE:b-rl2x03)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: be defensive about metric existing

Release note: None

couchand committed:
ui: fix ts issues from previous

Release note: None
